### PR TITLE
Switch to WASMI engine in icu_codepointtrie_builder

### LINKED ISSUE #3241: Update wasmer to 3.1.1
It turns out that `icu_codepointtrie_builder` depends on `wasmer` 2.2.1, which in its dependency tree contains the `mach` dependency. This dependency is no longer maintained, as it can be seen in the below advisory.

The good news is that `wasmer` 3.1.1 no longer depends on `mach`, so an update of the dependency should solve the issue.

----

> mach is unmaintained

| Details             |                                                |
| ------------------- | ---------------------------------------------- |
| Status              | unmaintained                |
| Package             | `mach`                      |
| Version             | `0.3.2`                   |
| URL                 | [https://github.com/fitzgen/mach/issues/63](https://github.com/fitzgen/mach/issues/63) |
| Date                | 2020-07-14                         |

Last release was almost 4 years ago.

Maintainer(s) seem to be completely unreachable. 

## Possible Alternative(s)

These may or may not be suitable alternatives and have not been vetted in any way;
- [mach2](https://crates.io/crates/mach2) - direct fork

See [advisory page](https://rustsec.org/advisories/RUSTSEC-2020-0168.html) for additional details.

### PR DESCRIPTION
**Pull request by sffc**
Fixes #3241 

I switched our WASM engine to `wasmi` because it is faster to compile and has fewer dependencies. I changed how we interact with C++, too, in order to improve the performance.

### CODE REVIEW

**Review by sffc — COMMENTED** [review]
The Cargo.lock file gets bigger with this change, even though I am still disabling most features of `wasmer` 😢

**Comment by sffc** [comment]
I also get errors like

> error: package `virtual-fs v0.11.0` cannot be built because it requires rustc 1.70 or newer, while the currently active rustc version is 1.67.1

which I need to debug, hopefully finding a compatible version of everything

**Comment by sffc** [comment]
Here is the output of depscheck with this PR:

```
Found two versions for `syn` (v1.0.109 & v2.0.48)
	Found non-allowlisted crate `ambient-authority`, consider adding to `EXTRA_DATAGEN_DEPS` or `EXTRA_ZIP_DEPS` or `EXTRA_RAYON_DEPS` in tools/depcheck/src/allowlist.rs if intentional
	Found non-allowlisted crate `anyhow`, consider adding to `EXTRA_DATAGEN_DEPS` or `EXTRA_ZIP_DEPS` or `EXTRA_RAYON_DEPS` in tools/depcheck/src/allowlist.rs if intentional
	Found non-allowlisted crate `async-trait`, consider adding to `EXTRA_DATAGEN_DEPS` or `EXTRA_ZIP_DEPS` or `EXTRA_RAYON_DEPS` in tools/depcheck/src/allowlist.rs if intentional
	Found non-allowlisted crate `bitflags`, consider adding to `EXTRA_DATAGEN_DEPS` or `EXTRA_ZIP_DEPS` or `EXTRA_RAYON_DEPS` in tools/depcheck/src/allowlist.rs if intentional
	Found non-allowlisted crate `cap-fs-ext`, consider adding to `EXTRA_DATAGEN_DEPS` or `EXTRA_ZIP_DEPS` or `EXTRA_RAYON_DEPS` in tools/depcheck/src/allowlist.rs if intentional
	Found non-allowlisted crate `cap-primitives`, consider adding to `EXTRA_DATAGEN_DEPS` or `EXTRA_ZIP_DEPS` or `EXTRA_RAYON_DEPS` in tools/depcheck/src/allowlist.rs if intentional
	Found non-allowlisted crate `cap-rand`, consider adding to `EXTRA_DATAGEN_DEPS` or `EXTRA_ZIP_DEPS` or `EXTRA_RAYON_DEPS` in tools/depcheck/src/allowlist.rs if intentional
	Found non-allowlisted crate `cap-std`, consider adding to `EXTRA_DATAGEN_DEPS` or `EXTRA_ZIP_DEPS` or `EXTRA_RAYON_DEPS` in tools/depcheck/src/allowlist.rs if intentional
	Found non-allowlisted crate `cap-time-ext`, consider adding to `EXTRA_DATAGEN_DEPS` or `EXTRA_ZIP_DEPS` or `EXTRA_RAYON_DEPS` in tools/depcheck/src/allowlist.rs if intentional
	Found non-allowlisted crate `dirs`, consider adding to `EXTRA_DATAGEN_DEPS` or `EXTRA_ZIP_DEPS` or `EXTRA_RAYON_DEPS` in tools/depcheck/src/allowlist.rs if intentional
	Found non-allowlisted crate `dirs-sys`, consider adding to `EXTRA_DATAGEN_DEPS` or `EXTRA_ZIP_DEPS` or `EXTRA_RAYON_DEPS` in tools/depcheck/src/allowlist.rs if intentional
	Found non-allowlisted crate `downcast-rs`, consider adding to `EXTRA_DATAGEN_DEPS` or `EXTRA_ZIP_DEPS` or `EXTRA_RAYON_DEPS` in tools/depcheck/src/allowlist.rs if intentional
	Found non-allowlisted crate `fs-set-times`, consider adding to `EXTRA_DATAGEN_DEPS` or `EXTRA_ZIP_DEPS` or `EXTRA_RAYON_DEPS` in tools/depcheck/src/allowlist.rs if intentional
	Found non-allowlisted crate `getrandom`, consider adding to `EXTRA_DATAGEN_DEPS` or `EXTRA_ZIP_DEPS` or `EXTRA_RAYON_DEPS` in tools/depcheck/src/allowlist.rs if intentional
	Found non-allowlisted crate `heck`, consider adding to `EXTRA_DATAGEN_DEPS` or `EXTRA_ZIP_DEPS` or `EXTRA_RAYON_DEPS` in tools/depcheck/src/allowlist.rs if intentional
	Found non-allowlisted crate `indexmap-nostd`, consider adding to `EXTRA_DATAGEN_DEPS` or `EXTRA_ZIP_DEPS` or `EXTRA_RAYON_DEPS` in tools/depcheck/src/allowlist.rs if intentional
	Found non-allowlisted crate `io-extras`, consider adding to `EXTRA_DATAGEN_DEPS` or `EXTRA_ZIP_DEPS` or `EXTRA_RAYON_DEPS` in tools/depcheck/src/allowlist.rs if intentional
	Found non-allowlisted crate `io-lifetimes`, consider adding to `EXTRA_DATAGEN_DEPS` or `EXTRA_ZIP_DEPS` or `EXTRA_RAYON_DEPS` in tools/depcheck/src/allowlist.rs if intentional
	Found non-allowlisted crate `ipnet`, consider adding to `EXTRA_DATAGEN_DEPS` or `EXTRA_ZIP_DEPS` or `EXTRA_RAYON_DEPS` in tools/depcheck/src/allowlist.rs if intentional
	Found non-allowlisted crate `is-terminal`, consider adding to `EXTRA_DATAGEN_DEPS` or `EXTRA_ZIP_DEPS` or `EXTRA_RAYON_DEPS` in tools/depcheck/src/allowlist.rs if intentional
	Found non-allowlisted crate `leb[128](https://github.com/unicode-org/icu4x/actions/runs/7929915708/job/21651128951?pr=4621#step:11:129)`, consider adding to `EXTRA_DATAGEN_DEPS` or `EXTRA_ZIP_DEPS` or `EXTRA_RAYON_DEPS` in tools/depcheck/src/allowlist.rs if intentional
	Found non-allowlisted crate `linux-raw-sys`, consider adding to `EXTRA_DATAGEN_DEPS` or `EXTRA_ZIP_DEPS` or `EXTRA_RAYON_DEPS` in tools/depcheck/src/allowlist.rs if intentional
	Found non-allowlisted crate `maybe-owned`, consider adding to `EXTRA_DATAGEN_DEPS` or `EXTRA_ZIP_DEPS` or `EXTRA_RAYON_DEPS` in tools/depcheck/src/allowlist.rs if intentional
	Found non-allowlisted crate `paste`, consider adding to `EXTRA_DATAGEN_DEPS` or `EXTRA_ZIP_DEPS` or `EXTRA_RAYON_DEPS` in tools/depcheck/src/allowlist.rs if intentional
	Found non-allowlisted crate `pin-project-lite`, consider adding to `EXTRA_DATAGEN_DEPS` or `EXTRA_ZIP_DEPS` or `EXTRA_RAYON_DEPS` in tools/depcheck/src/allowlist.rs if intentional
	Found non-allowlisted crate `ppv-lite86`, consider adding to `EXTRA_DATAGEN_DEPS` or `EXTRA_ZIP_DEPS` or `EXTRA_RAYON_DEPS` in tools/depcheck/src/allowlist.rs if intentional
	Found non-allowlisted crate `rand`, consider adding to `EXTRA_DATAGEN_DEPS` or `EXTRA_ZIP_DEPS` or `EXTRA_RAYON_DEPS` in tools/depcheck/src/allowlist.rs if intentional
	Found non-allowlisted crate `rand_chacha`, consider adding to `EXTRA_DATAGEN_DEPS` or `EXTRA_ZIP_DEPS` or `EXTRA_RAYON_DEPS` in tools/depcheck/src/allowlist.rs if intentional
	Found non-allowlisted crate `rand_core`, consider adding to `EXTRA_DATAGEN_DEPS` or `EXTRA_ZIP_DEPS` or `EXTRA_RAYON_DEPS` in tools/depcheck/src/allowlist.rs if intentional
	Found non-allowlisted crate `rustix`, consider adding to `EXTRA_DATAGEN_DEPS` or `EXTRA_ZIP_DEPS` or `EXTRA_RAYON_DEPS` in tools/depcheck/src/allowlist.rs if intentional
	Found non-allowlisted crate `shellexpand`, consider adding to `EXTRA_DATAGEN_DEPS` or `EXTRA_ZIP_DEPS` or `EXTRA_RAYON_DEPS` in tools/depcheck/src/allowlist.rs if intentional
	Found non-allowlisted crate `spin`, consider adding to `EXTRA_DATAGEN_DEPS` or `EXTRA_ZIP_DEPS` or `EXTRA_RAYON_DEPS` in tools/depcheck/src/allowlist.rs if intentional
	Found non-allowlisted crate `system-interface`, consider adding to `EXTRA_DATAGEN_DEPS` or `EXTRA_ZIP_DEPS` or `EXTRA_RAYON_DEPS` in tools/depcheck/src/allowlist.rs if intentional
	Found non-allowlisted crate `tracing`, consider adding to `EXTRA_DATAGEN_DEPS` or `EXTRA_ZIP_DEPS` or `EXTRA_RAYON_DEPS` in tools/depcheck/src/allowlist.rs if intentional
	Found non-allowlisted crate `tracing-attributes`, consider adding to `EXTRA_DATAGEN_DEPS` or `EXTRA_ZIP_DEPS` or `EXTRA_RAYON_DEPS` in tools/depcheck/src/allowlist.rs if intentional
	Found non-allowlisted crate `tracing-core`, consider adding to `EXTRA_DATAGEN_DEPS` or `EXTRA_ZIP_DEPS` or `EXTRA_RAYON_DEPS` in tools/depcheck/src/allowlist.rs if intentional
	Found non-allowlisted crate `wasi-cap-std-sync`, consider adding to `EXTRA_DATAGEN_DEPS` or `EXTRA_ZIP_DEPS` or `EXTRA_RAYON_DEPS` in tools/depcheck/src/allowlist.rs if intentional
	Found non-allowlisted crate `wasi-common`, consider adding to `EXTRA_DATAGEN_DEPS` or `EXTRA_ZIP_DEPS` or `EXTRA_RAYON_DEPS` in tools/depcheck/src/allowlist.rs if intentional
	Found non-allowlisted crate `wasmi`, consider adding to `EXTRA_DATAGEN_DEPS` or `EXTRA_ZIP_DEPS` or `EXTRA_RAYON_DEPS` in tools/depcheck/src/allowlist.rs if intentional
	Found non-allowlisted crate `wasmi_arena`, consider adding to `EXTRA_DATAGEN_DEPS` or `EXTRA_ZIP_DEPS` or `EXTRA_RAYON_DEPS` in tools/depcheck/src/allowlist.rs if intentional
	Found non-allowlisted crate `wasmi_core`, consider adding to `EXTRA_DATAGEN_DEPS` or `EXTRA_ZIP_DEPS` or `EXTRA_RAYON_DEPS` in tools/depcheck/src/allowlist.rs if intentional
	Found non-allowlisted crate `wasmi_wasi`, consider adding to `EXTRA_DATAGEN_DEPS` or `EXTRA_ZIP_DEPS` or `EXTRA_RAYON_DEPS` in tools/depcheck/src/allowlist.rs if intentional
	Found non-allowlisted crate `wasmparser-nostd`, consider adding to `EXTRA_DATAGEN_DEPS` or `EXTRA_ZIP_DEPS` or `EXTRA_RAYON_DEPS` in tools/depcheck/src/allowlist.rs if intentional
	Found non-allowlisted crate `wast`, consider adding to `EXTRA_DATAGEN_DEPS` or `EXTRA_ZIP_DEPS` or `EXTRA_RAYON_DEPS` in tools/depcheck/src/allowlist.rs if intentional
	Found non-allowlisted crate `wiggle`, consider adding to `EXTRA_DATAGEN_DEPS` or `EXTRA_ZIP_DEPS` or `EXTRA_RAYON_DEPS` in tools/depcheck/src/allowlist.rs if intentional
	Found non-allowlisted crate `wiggle-generate`, consider adding to `EXTRA_DATAGEN_DEPS` or `EXTRA_ZIP_DEPS` or `EXTRA_RAYON_DEPS` in tools/depcheck/src/allowlist.rs if intentional
	Found non-allowlisted crate `wiggle-macro`, consider adding to `EXTRA_DATAGEN_DEPS` or `EXTRA_ZIP_DEPS` or `EXTRA_RAYON_DEPS` in tools/depcheck/src/allowlist.rs if intentional
	Found non-allowlisted crate `witx`, consider adding to `EXTRA_DATAGEN_DEPS` or `EXTRA_ZIP_DEPS` or `EXTRA_RAYON_DEPS` in tools/depcheck/src/allowlist.rs if intentional
```

@Manishearth

**Comment by sffc** [comment]
On main:

```
INFO  [icu_datagen::driver] Generated key segmenter/sentence@1 (3.360s, flushed in 64.444ms)
INFO  [icu_datagen::driver] Generated key segmenter/word@1 (4.779s, flushed in 62.042ms)
INFO  [icu_datagen::driver] Generated key segmenter/grapheme@1 (4.805s, flushed in 55.470ms)
INFO  [icu_datagen::driver] Generated key segmenter/line@1 (9.127s, flushed in 79.752ms)
```

On this PR:

```
INFO  [icu_datagen::driver] Generated key segmenter/grapheme@1 (6.838s, flushed in 55.157ms)
INFO  [icu_datagen::driver] Generated key segmenter/sentence@1 (8.447s, flushed in 63.641ms)
INFO  [icu_datagen::driver] Generated key segmenter/word@1 (8.472s, flushed in 61.680ms)
INFO  [icu_datagen::driver] Generated key segmenter/line@1 (29.712s, flushed in 78.175ms)
```

Please let me know what you all think. I don't want to sink more time into this until we know which direction we're taking.

**Review by Manishearth — APPROVED** [review]
I'm happy with this approach. I'm okay with the slowdown, though I think we can potentially speed things up by building via ranges instead of codepoint by codepoint.

File: components/collections/codepointtrie_builder/Cargo.toml [hunk 1]
```diff
@@ -31,42 +31,19 @@ independent = true
...
-optional = true
-features = [
-    # We are running on the local system (as opposed to JavaScript)
-    "sys",
-    # We don't need real filesystem access, so use mem-fs
-    "mem-fs"
-]
+zerovec = { workspace = true }
+wasmi = "0.31.2"
+wasmi_wasi = "0.31.2"
```
Comment (robertbastian) [inline components/collections/codepointtrie_builder/Cargo.toml:46]: issue: why are these non-optional?

File: components/collections/codepointtrie_builder/cpp/README.md [hunk 2]
```diff
@@ -1,108 +1,23 @@
...
-
-```bash
-$ wasirun list_to_ucptrie.wasm
-Takes 3 positional arguments: default value, error value, and trie type
+$ make ICU4C_SOURCE=/path/to/icu4c/source ucptrie_wrap.wasm
 ```
 
 You can then copy the wasm file up one directory in order to update the version shipped with ICU4X.
-
-## Native Executable
```
Comment (robertbastian) [inline components/collections/codepointtrie_builder/cpp/README.md:42]: issue: why are you deleting the native instructions?

File: components/collections/codepointtrie_builder/src/lib.rs [hunk 3]
```diff
@@ -71,10 +71,11 @@
 #![cfg_attr(
     not(test),
     deny(
-        clippy::indexing_slicing,
-        clippy::unwrap_used,
-        clippy::expect_used,
-        clippy::panic,
+        // This Datagen-only builder code may panic when interacting with WASM
```
Comment (robertbastian) [inline components/collections/codepointtrie_builder/src/lib.rs:74]: nit: datagen is not really a concept in this crate

**Comment by sffc** [comment]
I managed to get rid of the WASI requirement at runtime entirely, which eliminates a _ton_ of dependencies. We still need WASI for building the C++ code, but the linker strips out all of the runtime WASI symbols. This is enabled by Reactor Modules which are a relatively recent change to the WASI spec.

File: components/collections/codepointtrie_builder/cpp/README.md (hunk 2 above)
Comment (sffc) [reply components/collections/codepointtrie_builder/cpp/README.md:42]: They are obsolete. Previously I was expecting that we would use the same executable as either a WASM+WASI executable or a native executable with the only difference being how we plug the stdin/stdout. However, the native bindings for this crate ended up treating ICU4C as a system library without needing to build a separate C++ executable. I have therefore changed this code from being an executable named `list_to_ucptrie` to being a library named `ucptrie_wrap`.

File: components/collections/codepointtrie_builder/Cargo.toml (hunk 1 above)
Comment (sffc) [reply components/collections/codepointtrie_builder/Cargo.toml]: Fixed

File: components/collections/codepointtrie_builder/src/lib.rs (hunk 3 above)
Comment (sffc) [reply components/collections/codepointtrie_builder/src/lib.rs:74]: Updated the docs

File: components/collections/codepointtrie_builder/cpp/Makefile [hunk 4]
```diff
@@ -3,34 +3,72 @@
...
-	mkdir -p icu4c_obj/wasm/common
-	mkdir -p icu4c_obj/wasm/tools/toolutil
-	$(WASICXX) -DTRUE=1 -c -flto -I$(ICU4C_SOURCE)/common -I$(ICU4C_SOURCE)/tools/toolutil $< -o $@
+wasm_obj/icu4c/%.o: $(ICU4C_SOURCE)/%.cpp
+	mkdir -p wasm_obj/icu4c/common
+	$(CXX) --target=wasm32-unknown-wasi \
+		-DTRUE=1 \
+		-DU_DISABLE_RENAMING=1 \
+		-c \
+		-flto \
```
Comment (robertbastian) [inline components/collections/codepointtrie_builder/cpp/Makefile:17]: nit: document these flags

File: components/collections/codepointtrie_builder/src/wasm.rs [hunk 5]
```diff
@@ -4,95 +4,280 @@
...
-        .expect("function exists");
-    let exit_result = start.call(&[]);
+    let trie_type = match builder.trie_type {
+        TrieType::Fast => 0,
+        TrieType::Small => 1,
+    };
+    let width = match core::mem::size_of::<T::ULE>() {
+        2 => 0, // UCPTRIE_VALUE_BITS_16
+        4 => 1, // UCPTRIE_VALUE_BITS_32
+        1 => 2, // UCPTRIE_VALUE_BITS_8
```
Comment (robertbastian) [inline components/collections/codepointtrie_builder/src/wasm.rs:239]: sigh

File: components/collections/codepointtrie_builder/ucptrie_wrap.wasm
Comment (robertbastian) [inline components/collections/codepointtrie_builder/ucptrie_wrap.wasm:1]: impressive size reduction!

**Review by robertbastian — APPROVED** [review]

**Review by robertbastian — APPROVED** [review]

**Review by Manishearth — APPROVED** [review]

**Review by Manishearth — APPROVED** [review]
also hooray we don't need to dedup syn anymore

either way i have most of the work done for that in https://github.com/wasmi-labs/wasmi/pull/941

**Pull request merged**